### PR TITLE
linux: avoid re-adding unsupported kernel branches (4.14) in this case in updater scripts

### PR DESCRIPTION
**Pull request by Ma27**
## Description of changes

As pointed out by @alyssais : https://github.com/NixOS/nixpkgs/pull/274275#issuecomment-1872198407

* Added a textfile indicating the oldest supported kernel branch (currently 4.19) and use that for the hardened updater: fd6404922513bded6ee3d3db11fe96724aa9892a
* Also take oldest supported branch into account when generating kernels-org.json: cdbffe4c58426b4e7adf8e4302ce4654bca6cd08
* Remove leftover 4.14 information.
* <del>Decided to also push the kernel updates that I made while testing the script :) </del> done in #278085

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/linux/kernel/min-kernel-branch.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+4.19
```
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/min-kernel-branch.txt:1]: We're never going to support custom kernels that are older than the oldest supported stable kernel, so why not just read the minimum version from the kernel.org JSON file?

That way, dropping an old kernel just means removing it from the JSON, which as we've seen is what people expect to be able to do anyway.

Comment (Ma27) [reply pkgs/os-specific/linux/kernel/min-kernel-branch.txt]: That's a good point, updated accordingly! :)

File: pkgs/os-specific/linux/kernel/hardened/update.py [hunk 2]
```diff
@@ -39,6 +39,16 @@
 })
 
 
+def read_min_kernel_branch():
+    with open(NIXPKGS_KERNEL_PATH / "kernels-org.json") as f:
+        components = list(map(
+            int,
+            sorted(json.loads(f.read()).keys())[0].split('.')
```
Comment (K900) [inline pkgs/os-specific/linux/kernel/hardened/update.py:46]: Can do `[int(x) for x in ...]` and `json.load(f)` here for readability. Also maybe use `packaging.Version` to parse the version specifiers? Not sure if it's worth the dependency, this line just scares me.

File: pkgs/os-specific/linux/kernel/update-mainline.py [hunk 3]
```diff
@@ -80,6 +81,26 @@ def get_hash(kernel: KernelRelease):
     return f"sha256:{hash}"
 
 
+def get_oldest_branch() -> List[int]:
+    with open(VERSIONS_FILE) as f:
+        components = list(map(
+            int,
+            sorted(json.loads(f.read()).keys())[0].split('.')
```
Comment (K900) [inline pkgs/os-specific/linux/kernel/update-mainline.py:88]: Same here.

**Pull request merged**